pkg/ingress/gateway.go

```diff
@@ -0,0 +1,203 @@
+//    updates/removes the certificate and secret as necessary.
+func (c client) provisionIngressGatewayCert(stop <-chan struct{}) error {
+	meshConfig := c.cfg.GetMeshConfig()
+	if meshConfig == nil {
+		return errors.New("MeshConfig cannot be nil")
+	}
+
+	defaultCertSpec := meshConfig.Spec.Certificate.IngressGateway
+	if defaultCertSpec != nil {
+		// Issue a certificate for the default certificate spec
+		if err := c.createAndStoreGatewayCert(*defaultCertSpec); err != nil {
+			return errors.Wrap(err, "Error provisioning default ingress gateway cert")
+		}
+	}
+
+	// Initialize a watcher to watch for CREATE/UPDATE/DELETE on the ingress gateway cert spec
+	go c.handleCertificateChange(defaultCertSpec, stop)
+
+	return nil
+}
+
+// createAndStoreGatewayCert creates a certificate for the given certificate spec and stores
+// it in the referenced k8s secret is the spec is valid.
+func (c client) createAndStoreGatewayCert(spec configv1alpha1.IngressGatewayCertSpec) error {
+	if len(spec.SubjectAltNames) == 0 {
+		return errors.New("Ingress gateway certificate spec must specify at least 1 SAN")
+	}
+
+	// Validate the validity duration
+	certValidityDuration, err := time.ParseDuration(spec.ValidityDuration)
+	if err != nil {
+		return errors.Wrapf(err, "Invalid cert duration '%s' specified", spec.ValidityDuration)
+	}
+
+	// Validate the secret ref
+	if spec.Secret.Name == "" || spec.Secret.Namespace == "" {
+		return errors.Errorf("Ingreaa gateway cert secret's name and namespace cannot be nil, got %s/%s", spec.Secret.Namespace, spec.Secret.Name)
+	}
+
+	// Issue a certificate
+	// OSM only support configuring a single SAN per cert, so pick the first one
+	certCN := certificate.CommonName(spec.SubjectAltNames[0])
+
+	issuedCert, err := c.certProvider.IssueCertificate(certCN, certValidityDuration)
+	if err != nil {
+		return errors.Wrapf(err, "Error issuing a certificate for ingress gateway")
+	}
+
+	// Store the certificate in the referenced secret
+	if err := c.storeCertInSecret(issuedCert, spec.Secret); err != nil {
+		return errors.Wrapf(err, "Error storing ingress gateway cert in secret %s/%s", spec.Secret.Namespace, spec.Secret.Name)
+	}
+
+	return nil
+}
+
+// storeCertInSecret stores the certificate in the specified k8s TLS secret
+func (c client) storeCertInSecret(cert certificate.Certificater, secret corev1.SecretReference) error {
+	secretData := map[string][]byte{
+		"ca.crt":  cert.GetIssuingCA(),
+		"tls.crt": cert.GetCertificateChain(),
+		"tls.key": cert.GetPrivateKey(),
+	}
+
+	sec := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secret.Name,
+			Namespace: secret.Namespace,
+		},
+		Type: corev1.SecretTypeTLS,
+		Data: secretData,
+	}
+
+	_, err := c.kubeClient.CoreV1().Secrets(secret.Namespace).Create(context.Background(), sec, metav1.CreateOptions{})
+	if apierrors.IsAlreadyExists(err) {
+		_, err = c.kubeClient.CoreV1().Secrets(secret.Namespace).Update(context.Background(), sec, metav1.UpdateOptions{})
```

Comment:
Not sure this applies, feel free to dismiss. 
Do multireplica osm require to sync this cert on all instances? i see each is issuing a cert, but only one will succeed in updating it. Does it matter?

Reply:
As long as the secret is created even by a single controller, it's good enough for this use case. I don't think we can synchronize such operations between multiple controllers. The only thing we could do better is wrt not logging errors when such things happen because multiple controllers tried to do the same thing. I'll follow up on this and address any misleading logs as I start testing the feature end to end.

Reply:
sgtm